src/v/config/constraints.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/convert.h"
+#include "config/from_string_view.h"
+#include "config/property.h"
+#include "json/json.h"
+#include "ssx/sformat.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <yaml-cpp/node/node.h>
+
+#include <optional>
+#include <string>
+
+namespace cluster {
+struct topic_configuration;
+} // namespace cluster
+
+namespace config {
+enum class constraint_type {
+    restrikt = 0,
+    clamp = 1,
+};
+
+std::string_view to_string_view(constraint_type type);
+
+template<>
+std::optional<constraint_type>
+from_string_view<constraint_type>(std::string_view sv);
+
+using constraint_enabled_t = ss::bool_class<struct constraint_enabled_tag>;
+
+namespace {
```

Comment:
Do not use unnamed namespaces in header files _clang-tidy_ [cert-dcl59-cpp](https://clang.llvm.org/extra/clang-tidy/checks/cert/dcl59-cpp.html)